diff -Naur ell-0.40/ell/cert.c ell-0.40.tpg/ell/cert.c
--- ell-0.40/ell/cert.c	2021-05-02 11:06:43.000000000 +0000
+++ ell-0.40.tpg/ell/cert.c	2021-05-07 12:53:26.952976086 +0000
@@ -437,6 +437,11 @@
 	return NULL;
 }
 
+static void cert_keyring_cleanup(struct l_keyring **p)
+{
+	l_keyring_free(*p);
+}
+
 #define RETURN_ERROR(msg, args...)	\
 	do {	\
 		if (error) {	\
@@ -451,7 +456,7 @@
 					const char **error)
 {
 	struct l_keyring *ca_ring = NULL;
-	_auto_(l_keyring_free) struct l_keyring *verify_ring = NULL;
+	L_AUTO_CLEANUP_VAR(struct l_keyring *, verify_ring, cert_keyring_cleanup) = NULL;
 	struct l_cert *cert;
 	struct l_key *prev_key = NULL;
 	int verified = 0;
diff -Naur ell-0.40/ell/tls.c ell-0.40.tpg/ell/tls.c
--- ell-0.40/ell/tls.c	2021-05-02 11:06:43.000000000 +0000
+++ ell-0.40.tpg/ell/tls.c	2021-05-07 12:54:28.972734292 +0000
@@ -1888,11 +1888,16 @@
 			"ServerHello decode error");
 }
 
+static void certchain_cleanup(struct l_certchain **p)
+{
+	l_certchain_free(*p);
+}
+
 static void tls_handle_certificate(struct l_tls *tls,
 					const uint8_t *buf, size_t len)
 {
 	size_t total;
-	_auto_(l_certchain_free) struct l_certchain *certchain = NULL;
+	L_AUTO_CLEANUP_VAR(struct l_certchain *, certchain, certchain_cleanup) = NULL;
 	struct l_cert *leaf;
 	size_t der_len;
 	const uint8_t *der;
diff -Naur ell-0.40/ell/useful.h ell-0.40.tpg/ell/useful.h
--- ell-0.40/ell/useful.h	2021-03-29 12:19:13.000000000 +0000
+++ ell-0.40.tpg/ell/useful.h	2021-05-07 12:52:06.863363194 +0000
@@ -52,13 +52,5 @@
 	return (oct >> start) & mask;
 }
 
-#define __AUTODESTRUCT(var, func)			\
-	void cleanup_ ## var(void *ptr)			\
-	{ func(*(void **) ptr); }			\
-	__attribute((cleanup(cleanup_ ## var)))
-
-#define _AUTODESTRUCT(var, func)			\
-	__AUTODESTRUCT(var, func)
-
-#define _auto_(func)					\
-	_AUTODESTRUCT(__COUNTER__, func)
+#define L_AUTO_CLEANUP_VAR(vartype,varname,destroy)	\
+       vartype varname __attribute__((cleanup(destroy)))
